src/main.mligo

```diff
@@ -0,0 +1,154 @@
+#import "ligo-generic-fa2/lib/multi_asset/fa2.mligo" "FA2"
+#import "./constants.mligo" "Constants"
+#import "./storage.mligo" "Storage"
+#import "./extension.mligo" "Extension"
+#import "./errors.mligo" "Errors"
+
+type storage = Storage.t
+type result = operation list * storage
+
+type create_token = {
+   token_id : nat;
+   data     : FA2.TokenMetadata.data;
+}
+
+type mint_or_burn = [@layout:comb] {
+   owner    : address;
+   token_id : nat;
+   amount_  : nat;
+}
+
+type permit_params = (key * (signature * bytes))
+type expiry_params = (address * (nat * (bytes option)))
+
+let create ({token_id;data},owner,amount : create_token * address * nat) (s : storage) =
+   let () = Extension.assert_admin s.extension in
+   let md = Storage.add_new_token s.token_metadata token_id data in
+   let s = Storage.set_token_metadata s md in
+   let ledger = FA2.Ledger.increase_token_amount_for_user s.ledger owner token_id amount in
+   let s = FA2.Storage.set_ledger s ledger in
+   ([]: operation list),s
+
+let mint (lst : mint_or_burn list) (s : storage) =
+   let () = Extension.assert_admin s.extension in
+   let process_one (ledger,{owner;token_id;amount_} : FA2.Ledger.t * mint_or_burn) =
+      let () = FA2.Storage.assert_token_exist  s token_id in
+      FA2.Ledger.increase_token_amount_for_user ledger owner token_id amount_
+   in
+   let ledger = List.fold_left process_one s.ledger lst in
+   let s = FA2.Storage.set_ledger s ledger in
+   ([]: operation list),s
+
+let burn (lst : mint_or_burn list) (s : storage) =
+   let () = Extension.assert_admin s.extension in
+   let process_one (ledger,{owner;token_id;amount_} : FA2.Ledger.t * mint_or_burn) =
+      FA2.Ledger.decrease_token_amount_for_user ledger owner token_id amount_
```

Comment:
For `burn` instead of decreasing the amount, would it make sense to have a burn address for each type of token, and make a transfer from owner to burn_address? (Not sure it makes sense, I feel this was as it will allow tracking of how many tokens were minted)

Reply:
oh yes thank you, this is a missing piece
to keep leveraging the contract-catalogue I'll add a total_supply field to the storage, similar to https://github.com/oxheadalpha/smart-contracts/blob/master/fa2_hooks/ligo/src/token_manager.mligo
I left this unresolved until it's done

Reply:
added a supply field to extension, and added missing tests for create_token, mint_token and burn_token entry points